Doc/library/itertools.rst

```diff
@@ -435,8 +435,13 @@ loops that truncate the stream.
           # islice('ABCDEFG', 2, 4) --> C D
           # islice('ABCDEFG', 2, None) --> C D E F G
           # islice('ABCDEFG', 0, None, 2) --> A C E G
+          # it = iter('abcdefghi')
+          # list(islice(it, 4, 4)) --> []
+          # list(it) --> ['e', 'f', 'g', 'h', 'i']
           s = slice(*args)
-          it = iter(range(s.start or 0, s.stop or sys.maxsize, s.step or 1))
+          it = iter(range(0, s.stop or sys.maxsize, s.step or 1))
+          for i in zip(range(0, s.start or 0), iterable):
```

Comment:
I'm thinking the additional code just covers a special case, so it would be better to put the new code inside the exception handler.  That way, it doesn't interfere with out understanding of the common case and it doesn't slow down the common case.  It also allows us to put in a clarifying comment about why the code is there.  Also, the variable on the for-loop should be "i, element" to make clear what the zip components are.  Trapping the tuple in a single pair named *i* is likely confusing because it means a tuple in one place and an index offset in other places:

        s = slice(*args)
        it = iter(range(s.start or 0, s.stop or sys.maxsize, s.step or 1))
        try:
            nexti = next(it)
        except StopIteration:
            # consume the iterable up the *start* position:
            for i, element in zip(range(s.start or 0), iterable):
                pass
            return
        for i, element in enumerate(iterable):
            if i == nexti:
                yield element
                nexti = next(it)